### PR TITLE
Some XProc for running XSpec

### PR DESCRIPTION
**Pull request by wendellpiez**
# Committer Notes

Addressing #93, this contains some XProc for running the XProc harness for XSpec in the neighbor `xspec` submodule.

Scripts and pipeline definition files (XProc) provide models that can be emulated.

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema-xslt/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema-xslt/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them? **readme is included**
- [x] Have you written new tests for your core changes, as applicable? **test directory is provided with test resources**
- [x] Have you included examples of how to use your new feature(s)? **ditto**

### CODE REVIEW

**Comment by wendellpiez** [comment]
Please ask me for clarification re: close and longer-term goals, non-goals and potentials of this spinoff. For now, this is meant mainly to extend the XSpec capabilities just a little bit more, in the context of other testing and integration.

This PR will help accelerate work on parallel efforts (#91, #92). The included HTML XSLT is over-engineered for the purpose, but a piece of work in itself (and functional standalone).

File: support/xspec-dev/xspec-assurance.sch
Comment (galtm) [inline support/xspec-dev/xspec-assurance.sch]: The assertion is good. Do you want to broaden the rule context to cover other XSpec elements that can have an `href` attribute? `x:expect`, `x:param`, `x:variable`, `x:import`

File: support/xspec-dev/testing/copy_me.xsl
Comment (galtm) [inline support/xspec-dev/testing/copy_me.xsl]: If you like to have a few namespace declarations as possible, you can remove these lines. Of course, they're not causing any harm if you leave them alone.

File: support/xspec-dev/testing/xspec-basic.xspec
Comment (galtm) [inline support/xspec-dev/testing/xspec-basic.xspec]: If you move line 14 so it's between 12 and 13, then you can remove lines 19, 23, 27, 31, and 35 because all the child scenarios will inherit the function call from the parent scenario.

File: support/xspec-dev/testing/xspec-basic.xspec
Comment (galtm) [inline support/xspec-dev/testing/xspec-basic.xspec]: Maybe you meant `../xspec-assurance.sch`. Same thing in xspec-shell.xspec.

File: support/xspec-dev/testing/xspec-shell.xspec
Comment (galtm) [inline support/xspec-dev/testing/xspec-shell.xspec]: This syntax accomplishes your purpose here but could lead to a confusing bug if the syntax is imitated in more substantive tests that you or someone else writes. I recommend changing `test="true()"` to `select="true()"` and making an analogous change to `select="false()"` on line 15. With those changes, you're saying, "check whether the function named in x:call returned true (resp., false)". Without those changes, you're saying "check whether the true() function returned true (resp., false)" **regardless of what x:call says** -- and that's the potential bug you'd want to avoid in a more substantive test.

BTW, I provided more explanation and alternatives at [Testing Boolean-Valued Functions](https://medium.com/@xspectacles/testing-boolean-valued-functions-xspec-blooper-4c27b2055f0b).

File: support/xspec-dev/xspec-summarize.xsl
Comment (galtm) [inline support/xspec-dev/xspec-summarize.xsl:39]: Is this part of what you might want to contribute to the xspec/xspec repo someday? If so, can this line generalize to cover tests for XQuery or Schematron? No action needed now, but maybe file the latter question away for later.

File: support/xspec-dev/xspec-mx-html-report.xsl
Comment (galtm) [inline support/xspec-dev/xspec-mx-html-report.xsl:264]: I really like that it is easy to copy content from the report! Please make this apparent, though, so a user doesn't click without that intention (e.g., to give focus to the browser application) and then find that they've lost something important that was on their clipboard. Many web sites use an icon, perhaps with a tooltip that says clicking will copy content, and then after you click, there is some indication like "Copied!"

File: support/xspec-dev/planning.md
Comment (galtm) [inline support/xspec-dev/planning.md]: I'd welcome your involvement in the XSpec project!

File: support/xspec-dev/xspec-mx-html-report.xsl
Comment (galtm) [inline support/xspec-dev/xspec-mx-html-report.xsl]: If the original test uses `<x:pending>` without a label, I think the result XML file will have `pending=""`, which doesn't match `\S`. There may be other parts of this XSLT that need some adjustment if you want to support `<x:pending>` without a label.

File: support/xspec-dev/xspec-mx-html-report.xsl
Comment (galtm) [inline support/xspec-dev/xspec-mx-html-report.xsl:227]: It's a matter of preference, but I found myself wishing for sibling x:expect items to appear in the report on separate lines instead of running together.

**Review by galtm — COMMENTED** [review]
Great work, @wendellpiez ! Here are my comments so far. Most are nitpicky and/or tangential. I didn't look at the .sh or .xpl files much, because I don't have Linux access and don't know XProc (yet, anyway). I did find a way to apply your xspec-mx-html-report.xsl to my own tests, and I'll try that out for a little while.

File: support/xspec-dev/xspec-assurance.sch
Comment (wendellpiez) [reply support/xspec-dev/xspec-assurance.sch]: Now matching `*[exists(@href)]`. Ha!

There is more to be done in this Schematron although I have not determined exactly what.

File: support/xspec-dev/testing/copy_me.xsl
Comment (wendellpiez) [reply support/xspec-dev/testing/copy_me.xsl]: Thanks for prompting more mindfulness! (Dumb machine.)

File: support/xspec-dev/testing/xspec-shell.xspec
Comment (wendellpiez) [reply support/xspec-dev/testing/xspec-shell.xspec]: Thanks, corrected and noted.

File: support/xspec-dev/xspec-summarize.xsl
Comment (wendellpiez) [reply support/xspec-dev/xspec-summarize.xsl:39]: Until I have a more definite plan I am sticking with a comment in the code noting the current limitation. (Which should also be noted elsewhere.)

This effort could also be folded into a general effort to align more systematically with the tagging and semantics of the XSpec report format, both as produced by tools and as an abstraction.

File: support/xspec-dev/xspec-mx-html-report.xsl
Comment (wendellpiez) [reply support/xspec-dev/xspec-mx-html-report.xsl:264]: Now considering how to do this with the most minimal JS and tagging overhead. Even without popups some simple things could likely be done to make implicit or explicit warnings, using CSS `:focus` and/or other selectors.

File: support/xspec-dev/xspec-mx-html-report.xsl
Comment (wendellpiez) [reply support/xspec-dev/xspec-mx-html-report.xsl]: Thanks! Habit learned from writing XSLTs for other kinds of source data ... we will have to return to `x:pending` when coming back to coverage.

**Comment by wendellpiez** [comment]
Soon to make a new commit addressing most @galtm comments so far thanks!

I am leaving one category, namely enhancements to the HTML presentation for usability (and indeed coverage). Not because they are uninteresting or unimportant, but mainly because dealing with this will be much easier in a code base with much more XSpec available to run and try out ... which we will shortly have **at least if we merge this PR sooner**, not later.

So a number of known issues are now on come-back-to:

* Nesting scenarios and reusing `x:context`
* Multiple `x:expect` together
* `x:label` instead of `@label`
* Copying functionality for code blocks should be Louder helping users not to overwrite their clipboard buffers by accident
* XQuery and Schematron report support

**Comment by wendellpiez** [comment]
Known issues and remaining to-dos are now noted on #96.

File: support/xspec-dev/readme.md
Comment (aj-stein-nist) [inline support/xspec-dev/readme.md]: Why is it not being merged in at this point? I am not a maintainer here just an occassional contributor and this question is based in a personal preference: the creation of multiple adjacent folders that outlive their intention makes a lot of the work harder to follow in the wrong run. In Metaschema (and previously OSCAL) projects, it has bitten me in the rear quite a few times.

File: support/xspec-dev/testing/xspec-test-batch.xpl
Comment (aj-stein-nist) [inline support/xspec-dev/testing/xspec-test-batch.xpl:26]: Blocking: I am apologize but there are a  lot of moving parts here and I don't quite understand if the issue is the `p:serialization` elements or how the shell script invokes them, but I consistently get an `XS0039` error. Can we review tomorrow? I am worried maybe I missed the obvious comment, but is this a TODO before final review and merge?

```sh
[com.xmlcalabash.drivers.Main.main()] ERROR com.xmlcalabash.util.DefaultXProcMessageListener - file:/home/me/code/metaschema-xslt/branches/wendell/issue93-xproc-for-xspec_A/support/xspec-dev/xspec-batch-report.xpl:10:27:err:XS0039:A p:serialization specifies a non-existant port.
[com.xmlcalabash.drivers.Main.main()] ERROR com.xmlcalabash.drivers.Main - It is a static error if the port specified on the p:serialization is not the name of an output port on the pipeline in which it appears or if more than one p:serialization element is applied to the same port.
```

File: support/xspec-dev/readme.md
Comment (aj-stein-nist) [inline support/xspec-dev/readme.md]: Non-blocking: it may help the uninitiated like me:

```suggestion
* Serialization needs to use latest (HTML5) serialization options to deal with character handling, rather than the current 'double reverse' character escaping of markup characters into [the Unicode Private Use Areas (PUA)](https://en.wikipedia.org/wiki/Private_Use_Areas) - thus effectively requiring a character mapping back, which is not well supported under XProc 1.0.
```

File: support/xspec-dev/readme.md
Comment (aj-stein-nist) [inline support/xspec-dev/readme.md:83]: Upon second review, it may help if you given more hand-holding here, as the first part of my review summarizing my trouble appears to be this but I am not sure reading the XSpec issue and these docs exactly what the relative import should be.

File: support/xspec-dev/testing/xspec-test-batch.xpl
Comment (wendellpiez) [reply support/xspec-dev/testing/xspec-test-batch.xpl:26]: This would have to be a bug in the XProc - will look. Maybe a port was twiddled and not re-checked....

File: support/xspec-dev/readme.md
Comment (wendellpiez) [reply support/xspec-dev/readme.md:83]: Fixing that bug would undoubtedly help a great deal.

This is why we do code review, thank you.

(Too bad we have no XSpec for XProc, ha. How to smoke-test an XProc run is an interesting question.)

File: support/xspec-dev/readme.md
Comment (wendellpiez) [reply support/xspec-dev/readme.md]: done in latest commit

File: support/xspec-dev/readme.md
Comment (wendellpiez) [reply support/xspec-dev/readme.md]: Fixed, thanks. (The readme was not clear that the neighbor folder is actually a completely discrete submodule and not under our maintenance.)

File: support/xspec-dev/testing/xspec-test-batch.xpl
Comment (aj-stein-nist) [reply support/xspec-dev/testing/xspec-test-batch.xpl:26]: OK, we can look together if you would like. I spent some time looking and even thought there is a specific error code (based in spec I assume with `XS0039`) I found exactly one mailing list thread with those encountering the issue and playing with different `p:serialization` declarations in both XPL files and could not get around it. I assume it is something obvious. Let me know how I can be helped?

https://www.mail-archive.com/xproc-dev@w3.org/msg00130.html

File: support/xspec-dev/xspec-batch-report.xpl
Comment (wendellpiez) [inline support/xspec-dev/xspec-batch-report.xpl:1]: @aj-stein-nist the glitch was in this file - discrepancy btw `x:serialization/@port` on line 24 and `p:output/@port` on line 25. The XProc processor complains when it sees a serialization setting that applies nowhere.

IME port names often wiggle towards the end as interfaces require improvement. In this case, I was sitting on the fence regarding whether that port is for 'internal' or 'external' use, or even whether it is needed (since it can also be dropped).

Thanks for patience!

Comment (aj-stein-nist) [reply support/xspec-dev/xspec-batch-report.xpl:1]: What's odd is I recall trying this fix because it seemed obvious and I still got an error, perhaps I did it in one shell script (`support/xspec-dev/testing/mvn-xproc-xspec-test-batch.sh`) and not the other? I don't know, I distinctly thought this would fix it but I must have done it wrong or I would have recommended this fix. Thanks for the help!

File: support/xspec-dev/mvn-xproc-xspec-html.sh
Comment (aj-stein-nist) [inline support/xspec-dev/mvn-xproc-xspec-html.sh:26]: Is there a reason that the output goes into a path relative to the shell script that is with a hard-coded prefix? If you use it outside the presumed directory structure and if the xspec folder fails it will fail if you do not create it (the current shell script does not). I would just put it in the adjacent directory and switch out the extension.

```suggestion
RESULT_FILE="${XSPEC_FILE%%.*}.html"
```

As a dev user, I also presume I am not unique in that I would run the command and expect to see the file, _as a default_ adjacent to the current folder in which I am operating from (so wherever the target XSpec file is). That did not happen when I was re-evaluating the updated branch for this PR, and had to run the shell script again quickly in debug mode to understand where the file was. This violates POLA/POLS, or the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

**Review by aj-stein-nist — APPROVED** [review]
I have one minor recommendation about the usability of the tool and I gave some detailed feedback as to why, but I think this is good and useful. I have some follow-up questions and concerns about how this work will span this repo, ixml-breadboard, and a variety of other projects.

I would say an important next step is being very intentional with organization and documenting how and why it will be set up that way across those respective repositories, as I do not have an obvious upfront answer as to how I can reuse or share this work across them without duplicatively copy-pasting, and that leads to a lot of headache in maintenance over the long haul (this is the key thing I learned from the OSCAL project). I hope that feedback is constructive.

File: support/xspec-dev/mvn-xproc-xspec-html.sh
Comment (wendellpiez) [reply support/xspec-dev/mvn-xproc-xspec-html.sh:26]: Now `{basename}-result.html` cf https://github.com/usnistgov/metaschema-xslt/pull/94/commits/4b2b86197f2b62a12e55b036607bf0e5378f694f

**Pull request merged**
